### PR TITLE
Concrete memory backing, unicorn page caching fix

### PR DESCRIPTION
**Pull request by Lukas-Dresel**
Implemented support for storing bytes in the memory model directly to speed up memory syncing between unicorn and angr directly. 

Reimplemented page cache syncing interface between angr and angr-native so that pages get accurately mapped into the unicorn cache. this fixes a bug where pages were in the cache that were not updated correctly.

### CODE REVIEW

**Comment by Lukas-Dresel** [comment]
I'm debating making this nicer now, with actually special cased `SimMemoryObject` instead of this if-else hacky stuff. What do you guys think?

**Comment by Lukas-Dresel** [comment]
As of right now I can't really think of other good types as MemoryObjects other than maybe `Clemory`s? So it might be over-engineering to handle until we need it.

**Comment by Lukas-Dresel** [comment]
Currently all data that is initialized to zero and from backers is stored as bytes, concrete stores currently are not. This is possible, but i'd say probably not very important? not sure

File: angr/storage/memory_object.py [hunk 1]
```diff
@@ -29,27 +37,49 @@ def last_addr(self):
...
+        if self.is_bytes != other.is_bytes:
+            return False
+
+        if self.is_bytes:
+            return self.object == other.object
+        else:
+            return self.object.cache_key == other.object.cache_key
+
+    def _length_equals(self, other):
+        if type(self.length) != other.length:
```
Comment (ltfish) [inline angr/storage/memory_object.py]: Why are we comparing a type to a `length` here?

File: angr/storage/memory_object.py [hunk 2]
```diff
@@ -9,12 +13,16 @@ class SimMemoryObject(object):
     SimSymbolicMemory class.
     """
     def __init__(self, obj, base, length=None, byte_width=8):
-        if not isinstance(obj, claripy.ast.Base):
+        if type(obj) == bytes:
```
Comment (ltfish) [inline angr/storage/memory_object.py]: `type(obj) is bytes`

File: angr/storage/memory_object.py [hunk 3]
```diff
@@ -1,6 +1,10 @@
 import claripy
 from ..errors import SimMemoryError
 
+def obj_bit_size(o):
+    if type(o) == bytes:
```
Comment (ltfish) [inline angr/storage/memory_object.py]: `type(o) is bytes`

File: angr/storage/paged_memory.py [hunk 4]
```diff
@@ -346,6 +346,31 @@ def __init__(self, memory_backer=None, permissions_backer=None, pages=None, init
         self._hash_mapping = cooldict.BranchingDict() if hash_mapping is None else hash_mapping
         self._updated_mappings = set()
 
+    def _page_align_down(self, x):
+        return x - (x % self._page_size)
+
+    def _page_align_up(self, x):
+        return self._page_align_down(x + self._page_size - 1)
+
+    def __is_page_aligned(self, x):
```
Comment (ltfish) [inline angr/storage/paged_memory.py]: Why double underscores?

File: angr/storage/paged_memory.py [hunk 5]
```diff
@@ -549,12 +579,14 @@ def _initialize_page(self, n, new_page):
...
-                        backer = self._memory_backer[i]
-                    elif isinstance(self._memory_backer[i], bytes):
-                        backer = claripy.BVV(self._memory_backer[i])
-                    else:
+                    backer = self._memory_backer[i]
+
+                    if not isinstance(self._memory_backer[i], (claripy.ast.Base, bytes)):
                         backer = claripy.BVV(self._memory_backer[i], self.byte_width)
+
+                    if type(backer) == bytes and self.byte_width != 8:
```
Comment (ltfish) [inline angr/storage/paged_memory.py]: `type(backer) is bytes`

**Comment by ltfish** [comment]
> Currently all data that is initialized to zero and from backers is stored as bytes, concrete stores currently are not. This is possible, but i'd say probably not very important? not sure

I'll implement it.

**Comment by ltfish** [comment]
Gentle ping? @Lukas-Dresel

**Comment by Lukas-Dresel** [comment]
Implemented the requested changes @ltfish

**Comment by ltfish** [comment]
I'll merge it in as soon as tests pass. Thanks!

**Pull request merged**